ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,777 @@
+ * become eligible to be fetched.
+ *
+ * This class implements {@link ReadableStreamChannel} and passes itself as the result of this operation.
+ */
+class GetBlobOperation extends GetOperation<ReadableStreamChannel> implements ReadableStreamChannel {
+  // the callback to use to complete the operation.
+  private final OperationCompleteCallback operationCompleteCallback;
+  // whether the operationCompleteCallback has been called already.
+  private final AtomicBoolean operationCallbackInvoked = new AtomicBoolean(false);
+  // The first chunk may be a metadata chunk if the blob is composite, or the only data chunk if the blob is simple.
+  private final FirstGetChunk firstChunk;
+  // Associated with all data chunks in the case of composite blobs. Only a fixed number of these are initialized.
+  // Each of these is initialized with the information required to fetch a data chunk and is responsible for
+  // retrieving and adding it to the list of chunk buffers. Once complete, they are reused to fetch subsequent data
+  // chunks.
+  private GetChunk[] dataChunks;
+  // the factory to use to deserialize keys in a metadata chunk.
+  private final BlobIdFactory blobIdFactory;
+  // the total number of data chunks associated with this blob.
+  private int numChunksTotal;
+  // the total number of data chunks retrieved so far (and may or may not have been written out yet).
+  private int numChunksRetrieved;
+  // a list iterator to the chunk ids that need to be fetched for this operation, if this is a composite blob.
+  private ListIterator<StoreKey> chunkIdIterator;
+  // chunk index to retrieved chunk buffer mapping.
+  private Map<Integer, ByteBuffer> chunkIndexToBuffer;
+  // To find the GetChunk to hand over the response quickly.
+  private final Map<Integer, GetChunk> correlationIdToGetChunk = new HashMap<>();
+
+  // ReadableStreamChannel implementation related variables:
+
+  // whether this ReadableStreamChannel is open.
+  private boolean isOpen = true;
+  // whether readInto() has been called yet by the caller on this ReadableStreamChannel.
+  private volatile boolean readCalled = false;
+  // The channel to write chunks of the blob into. This will be initialized when the caller calls the readInto().
+  private AsyncWritableChannel asyncWritableChannel;
+  // the callback to call when all the chunks are successfully written out into the asyncWritableChannel.
+  private Callback<Long> readIntoCallback;
+  // the future to mark as done when all the chunks are successfully written out into the asyncWritableChannel.
+  private FutureResult<Long> readIntoFuture;
+  // the number of bytes written out to the asyncWritableChannel. This would be the size of the blob eventually.
+  private Long bytesWritten = 0L;
+  // the number of chunks that have been written out to the asyncWritableChannel.
+  private final AtomicInteger numChunksWrittenOut = new AtomicInteger(0);
```

Comment:
`indexOfNextChunkToWriteOut` and most other variables are only accessed int the context of the `RequestResponseHandler` thread, so they don't have to be atomic or volatile.

`numChunksWrittenOut` is accessed both by the `asyncWritableChannel` callback invoker thread and the `RequestResponseHandler` thread. Additionally, we need an ordering guarantee due to the following scenario:

`RequestResponseHandler` thread does this (simplified):
```java
if (numChunksWrittenOut == numChunksTotal) {
  call readIntoCallback with bytesWritten
}
```
and the `asyncWritableChannel`' callback invoker does this:
```java
void onComplete(...) {
  bytesWritten += what_was_written;
  numChunksWrittenOut++;
}
```
which means that `numChunksWrittenOut` should be incremented only *after* `bytesWritten` is updated. So it has to be atomic or volatile. Just made it atomic due to the increment operation (volatile is probably not the right choice).